sound/soc/codecs/rt711-sdca.c

```diff
@@ -0,0 +1,1382 @@
+		case 0x08:
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+	}
+
+	/* Host is owner, so set back to device */
+	if (owner == 0) {
+		/* set owener to device */
+		regmap_write(rt711->regmap,
+			RT711_SDCA_CTL(FUN_HID, ENT_HID01,
+			CTL_HIDTX_SET_OWNER_TO_DEVICE, 0), 0x01);
+	}
+
+_end_btn_det_:
+	return btn_type;
+}
+
+static int rt711_sdca_headset_detect(struct rt711_priv *rt711)
+{
+	unsigned int det_mode;
+	int ret;
+
+	/* get detected_mode */
+	ret = regmap_read(rt711->regmap,
+		RT711_SDCA_CTL(FUN_JACK_CODEC, ENT_GE49,
+		CTL_DETECTED_MODE, 0), &det_mode);
+	if (ret < 0)
+		goto io_error;
+
+	switch (det_mode) {
+	case 0x00:
+		rt711->jack_type = 0;
+		break;
+	case 0x03:
+		rt711->jack_type = SND_JACK_HEADPHONE;
+		break;
+	case 0x05:
+		rt711->jack_type = SND_JACK_HEADSET;
+		break;
+	default:
+		break;
+	}
+
+	/* write selected_mode */
+	ret = regmap_write(rt711->regmap,
+		RT711_SDCA_CTL(FUN_JACK_CODEC, ENT_GE49,
+		CTL_SELECTED_MODE, 0), det_mode);
+	if (ret < 0)
+		goto io_error;
+
+	dev_dbg(&rt711->slave->dev,
+		"%s, detected_mode=0x%x\n", __func__, det_mode);
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static void rt711_sdca_jack_detect_handler(struct work_struct *work)
+{
+	struct rt711_priv *rt711 =
+		container_of(work, struct rt711_priv, jack_detect_work.work);
+	int btn_type = 0, ret;
+	unsigned int jack_status = 0, check_bit;
+
+	if (!rt711->hs_jack)
+		return;
+
+	if (!rt711->component->card->instantiated)
+		return;
+
+	ret = regmap_read(rt711->regmap,
+		RT711_SDCA_CTL(FUN_JACK_CODEC, ENT_GE49,
+		CTL_DETECTED_MODE, 0), &jack_status);
+	if (ret < 0)
+		goto io_error;
+	check_bit = BIT(0);
+
+	/* pin attached */
+	if (jack_status & check_bit) {
+		/* jack in */
+		if (rt711->jack_type == 0) {
+			ret = rt711_sdca_headset_detect(rt711);
+			if (ret < 0)
+				return;
+			if (rt711->jack_type == SND_JACK_HEADSET)
+				btn_type = rt711_sdca_button_detect(rt711);
+		} else if (rt711->jack_type == SND_JACK_HEADSET) {
+			/* jack is already in, report button event */
+			btn_type = rt711_sdca_button_detect(rt711);
+		}
+	} else {
+		/* jack out */
+		rt711->jack_type = 0;
+	}
+
+	dev_dbg(&rt711->slave->dev,
+		"in %s, jack_type=0x%x\n", __func__, rt711->jack_type);
+	dev_dbg(&rt711->slave->dev,
+		"in %s, btn_type=0x%x\n", __func__, btn_type);
+
+	snd_soc_jack_report(rt711->hs_jack, rt711->jack_type | btn_type,
+			SND_JACK_HEADSET |
+			SND_JACK_BTN_0 | SND_JACK_BTN_1 |
+			SND_JACK_BTN_2 | SND_JACK_BTN_3);
+
+	return;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+}
+
+static void rt711_sdca_jack_init(struct rt711_priv *rt711)
+{
+	struct sdw_slave_prop *prop = &rt711->slave->prop;
+
+	mutex_lock(&rt711->calibrate_mutex);
+
+	if (rt711->hs_jack) {
+		/* Enable HID1 event & set button RTC as OSC */
+		rt711_sdca_index_update_bits(rt711, RT711_VENDOR_HDA_CTL,
+			RT711_PUSH_BTN_INT_CTL6,
+			0x80f0, 0x80f0);
```

Comment:
``RT711_PUSH_BTN_INT_CTL6, 0x80f0, 0x80f0);`` is more readable. Same as the following lines.